azext_iot/monitor/property.py

```diff
@@ -7,91 +7,202 @@
+            )
+            > 1
+        )
+        return value
+
+    def _add_central_issue(self, severity: Severity, details: str):
+        self._issues_handler.add_central_issue(
+            severity=severity,
+            details=details,
+            message=None,
+            device_id=self._device_id,
+            template_id=self._template.id,
+        )
 
-        time.sleep(polling_interval_seconds)
+    def _is_interface(self, interface_name):
+        # Remove PNP interface prefix to get the actual interface name
+        interface_name = interface_name.replace(PNP_INTERFACE_PREFIX, "")
+        return interface_name in self._template.interfaces
 
-        prev_twin = twin
+    def _get_device_template(self):
 
+        central_device_provider = CentralDeviceProvider(self._cmd, self._app_id)
+        device = central_device_provider.get_device(self._device_id)
+        provider = CentralDeviceTemplateProvider(cmd=self._cmd, app_id=self._app_id)
+        template = provider.get_device_template(
+            device_template_id=device.instance_of,
+            central_dns_suffix=self._central_dns_suffix,
+        )
+        return template
 
-def compare_properties(prev_prop: Property, prop: Property):
-    if prev_prop.version == prop.version:
-        return
+    def start_property_monitor(self,):
+        prev_twin = None
 
-    changes = {
-        key: changed_props(prop.props[key], prop.metadata[key], key)
-        for key, val in prop.metadata.items()
-        if is_relevant(key, val)
-    }
+        device_twin_provider = CentralDeviceTwinProvider(
+            cmd=self._cmd, app_id=self._app_id, device_id=self._device_id
+        )
 
-    return changes
+        while True:
 
+            raw_twin = device_twin_provider.get_device_twin(
+                central_dns_suffix=self._central_dns_suffix
+            )
 
-def is_relevant(key, val):
-    if key in {"$lastUpdated", "$lastUpdatedVersion"}:
-        return False
+            twin = DeviceTwin(raw_twin)
+            if prev_twin:
+                change_d = self._compare_properties(
+                    prev_twin.desired_property, twin.desired_property,
+                )
+                change_r = self._compare_properties(
+                    prev_twin.reported_property, twin.reported_property
+                )
+
+                if change_d:
+                    print("Changes in desired properties:")
+                    print("version :", twin.desired_property.version)
+                    print(change_d)
+
+                if change_r:
+                    print("Changes in reported properties:")
+                    print("version :", twin.reported_property.version)
+                    print(change_r)
+            time.sleep(DEVICETWIN_POLLING_INTERVAL_SEC)
+
+            prev_twin = twin
+
+    def start_validate_property_monitor(self, minimum_severity):
+        prev_twin = None
+
+        device_twin_provider = CentralDeviceTwinProvider(
```

Comment:
maintenance/pattern: avoid creating providers in functions, try to create them in constructor

Reply:
ideally - don't construct other providers in providers, instead have a controller construct them and pass info as required